### ISSUE TITLE
Error in _escape_sf when attempting to render docs without publishing to Confluence

### ISSUE THREAD

**Issue by ncoghlan** [sphinx-contrib/confluencebuilder#235]
I'm attempting to test confluencebuilder compatibility with the various blockdiag plugins before setting up full publication support, but I'm getting an exception that appears to be related to working out the eventual publication URLs for the rendered images.

Last part of the traceback:

```
  File ".../site-packages/docutils/nodes.py", line 1912, in dispatch_visit
    return method(node)
  File ".../site-packages/sphinxcontrib/confluencebuilder/translator.py", line 1146, in visit_image
    hosting_doctitle = self._escape_sf(hosting_doctitle)
  File ".../site-packages/sphinxcontrib/confluencebuilder/translator.py", line 1953, in _escape_sf
    data = data.replace('&', '&amp;')
AttributeError: 'NoneType' object has no attribute 'replace'
```

Current configuration settings:

```
confluence_publish = False
confluence_server_url = "https://wiki.ourintranet.example"
confluence_space_name = "OurDepartment"
confluence_parent_page = "OurProjectDocs"
confluence_page_heirarchy = True
confluence_master_homepage = False
```

Before I dig into the issue too much further, I wanted to check if what I'm doing is something you would expect to work.

**Comment by jdknight** [comment]
The management of assets (images, downloads, etc.) is done to help track which document assets will be attached to (since an attachment may be attached to the associated document or the root document if duplicate assets are detected). The process is as follows:

1. When each document is built, it will be [processed for asset tracking](https://github.com/sphinx-contrib/confluencebuilder/blob/v1.1/sphinxcontrib/confluencebuilder/builder.py#L246)*.
2. When a doctree is resolved, it will be [processed again known missing assets](https://github.com/sphinx-contrib/confluencebuilder/blob/v1.1/sphinxcontrib/confluencebuilder/__init__.py#L32-L38) to help deal with extensions that generate images based off the document's content (e.g. `sphinx.ext.imgconverter`).
3. When building an image attachment, the implementation to [find the target document name](https://github.com/sphinx-contrib/confluencebuilder/blob/v1.1/sphinxcontrib/confluencebuilder/translator.py#L1144) which will be hosting the asset and reference in into the attachment page reference (if needed).

In the example case shown in the original post, it looks like `hosting_doctitle` is `None`. which in theory should not happen (although there could be a bug with something else); but given the context of this issue, I would suspect that there is another extension in play which is generate images between the `doctree-resolved` event and when the translator is being processed. One thing to consider is that this extension may be generating asset information too soon (i.e. there maybe a different event than `doctree-resolved` that should be used to handling pre-translation processing like `env-updated`; but I am unaware of what is the right solution at this time). If an extension is populating assets to be generated into a final document after a `doctree-resolved` event, I would be curious to see this in action (mainly, to improve my understanding of the Sphinx engine).

\* Now thinking about this step, this could be incorrect (although I would suspect not related to error mention above) since this may pre-process assets which may be later removed by other extensions before the writing stage.

On a relate note, see pending changes to this code in #228 which will change the `ConfluenceAssetManager` implementation to correct a separate issue, if you are thinking of changing anything (also #217 as well).

**Comment by ncoghlan** [comment]
OK, thanks for the extra info.

Given that, my next step was to retry the build with all the diagram nodes commented out. That made the error go away, and restoring a single `blockdiag` node brought it back, so it looks like this is indeed a rendering conflict with `sphinxcontrib-blockdiag` (et al).

The Sphinx integration code for that is in https://github.com/blockdiag/sphinxcontrib-blockdiag/blob/master/sphinxcontrib/blockdiag.py, and it looks like for non-html builders, the plugin does indeed use the `doctree-resolved` event to generate `PNG` files for all the diagrams, which means it would run *after* `confluencebuilder`'s own hook for that event given the order in which I was enabling extensions.

I reworked the `conf.py` file to change the order in which hooks get run (by putting `confluencebuilder` last in the list), and sure enough, that changed the error reported:

```
  File ".../site-packages/sphinxcontrib/confluencebuilder/util.py", line 35, in hashAsset
    with open(asset, 'rb') as file:
FileNotFoundError: [Errno 2] No such file or directory: '.../docs/blockdiag-daf9eeecb8d631c3baf5a975f7a4aecc4875bfd4.png'
```

**Comment by jdknight** [comment]
@ncoghlan, thanks for trying that out. I'll note for later to see if `env-updated` (or some other event) is a better event for this extension to use.

You may want to try the [`feature/graceful-asset-names`](https://github.com/sphinx-contrib/confluencebuilder/tree/feature/graceful-asset-names) branch (if interested) which does re-work some of the asset management. Although I think the logic still determines the absolute path in the same manner so it may not help. On that note, maybe just a bug in how the path is determined (I recall seeing some oddities in the past but other issues were pressing at the time).

**Comment by jdknight** [comment]
This issue was observed at a time where there were a combination of issues with the assert management implementation and some recent implementation issues where titleless documents with figures would break the document writing stage. With asset management changes already on the main branch and recent changes introduced in #374, there implementation should prevent the observed error conditions above (see also #315 for mentioned support for blockdiag-based directives). These changes will be available next stable release (v1.3+). Will mark as pending-next-release and will close up when the release has been completed (see #362 for more details).

**Comment by jdknight** [comment]
v1.3 is now available on PyPI -- marking as closed.
